README.md
# 优学院看视频-YouXueYuan_Watch-Video

* 可用来看优学院视频而不用手动点击
* 答题请手动，项目没打算增加答题功能  
  
项目地址  
[GitHub](https://github.com/Brush-JIM/YouXueYuan_Watch-Video)、[Bitbucket](https://bitbucket.org/Brush-JIM/youxueyuan-javascript/)、[Greasy Fork](https://greasyfork.org/zh-CN/scripts/382033-优学院-youxueyuan-看视频)  
  
Python程序版请到隔壁  
[GitHub](https://github.com/Brush-JIM/YouXueYuan-Python)、[Bitbucket](https://bitbucket.org/Brush-JIM/youxueyuan-python/)  
  
## 食用方法  
* 安装脚本管理器，方法请看[Greasy Fork](https://greasyfork.org/zh-CN)
* 访问  [https://greasyfork.org/zh-CN/scripts/382033-优学院-YouXueYuan-看视频](https://greasyfork.org/zh-CN/scripts/382033-优学院-YouXueYuan-看视频)
* 安装脚本到脚本管理器中  
* 进入课程，右下角有按钮  

# 日志
CHANGELOG.md [GitHub](https://github.com/Brush-JIM/YouXueYuan_Watch-Video/blob/master/CHANGELOG.md) or [Bitbucket](https://bitbucket.org/Brush-JIM/youxueyuan-javascript/src/master/CHANGELOG.md?fileviewer=file-view-default)  
  
---
代码开源、免费  
不接受任何捐赠  


CHANGELOG.md
# 日志  

## 19.11.30
* 重构代码，解决视频可能没看完就跳过的问题，取消公告栏

## 19.06.01
* 优化代码

## 19.5.30
* Fix a Bug.

## 19.5.28
* 修复一处导致脚本停止继续执行的问题

## 19.5.27
* 项目更名

## 19.5.26
#### 从19.5.21开始，开始重写，终于写完
* 重构脚本
* 改变数据保存方式
* 增加公告栏，Greasemonkey可能用不了
* 增强脚本稳定性

## 19.5.17
* 可能修复了小测之后下一页的视频可能跳过的问题
* 感谢@kuailezhiyuan[反馈](https://github.com/Brush-JIM/YouXueYuan_Watch-Video/issues/1)该问题

## 19.5.15
* 改善运行方式
* 增加完成课程后是否返回课程目录选项

## 19.5.2
* 修复一个Bug

## 19.5.1
* 改变关闭走神方法，基本实现运行脚本后永久关闭

## 19.4.30 V2.0
* 重构脚本
* 使用更多方法实现功能

## 19.4.23 V1.0
* 适配多视频播放
* 改变运作方式
* 可保存调速、音量信息
* 改变按钮位置
* 对Greasemonkey进行兼容

## 19.4.19 V0.4
* 去除开小差提示框  
* 增加速率保存功能

## 19.4.5
* 重新上传项目  

## 19.3.30 V0.3
* 增加调速功能

## 19.3.25 V0.2
* 优化代码  

## 19.3.23 V0.1
* 上传第一份  


Historic Version/1.1.user.js
// ==UserScript==
// @name         优学院-YouXueYuan-看视频
// @namespace    https://github.com/Brush-JIM/YouXueYuan-JavaScript
// @version      1.1
// @description  js没学过，技术纯属渣渣。可用来看优学院视频而不用手动点击。
// @author       Brush-JIM
// @match        https://ua.ulearning.cn/learnCourse/learnCourse.html?courseId=*&chapterId=*
// @require      https://code.jquery.com/jquery-3.3.1.min.js
// @grant        unsafeWindow
// @run-at       document-start
// @icon         https://www.ulearning.cn/ulearning/favicon.ico
// ==/UserScript==

(function() {
    'use strict';
    //删除统计元素
    document.getElementById("statModal").parentNode.removeChild(document.getElementById("statModal"));
    //删除指导元素
    document.getElementsByClassName("user-guide")[0].parentNode.removeChild(document.getElementsByClassName("user-guide")[0]);
    //侧边栏
    var cbl = document.createElement("div");
    cbl.setAttribute("style", "width: 60px;height: 74px;overflow: hidden;position: fixed;right: 10px;bottom: 10%;padding: 4px;background-color: #06a9f4;z-index: 9999;/* border-radius: 8px 0px 0px 8px; *//* box-shadow: rgba(0, 85, 255, 0.098) 0px 0px 20px 0px; */border: 1px solid rgb(233, 234, 236);");
    //cbl.setAttribute('style',"display:table-cell;*display:inline-block;*float:left;width:60px;height:70px;vertical-align:middle;border:1px solid #e3e3e9;margin:0;background-color:#fff;");
    cbl.setAttribute("id","set-auto");
    cbl.innerHTML='<div style="cursor: pointer; text-align: center; padding: 0px;"><span id="set-auto" style="font-size: 12px;line-height: 16px;color: rgb(255, 255, 255);">设置<br />&<br />开关</span></div>';
    document.querySelector('body').appendChild(cbl);
    var xfk = document.createElement('div');
    xfk.setAttribute("style", "position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px; display: none;");
    xfk.setAttribute("id", "set-mune-hide");
    xfk.innerHTML = '<div style="display: block;overflow: hidden;height: 300px;width: 300px;/* border-radius: 8px; *//* box-shadow: rgba(106, 115, 133, 0.22) 0px 6px 12px 0px; */border: 1px solid rgb(233, 234, 236);background-color: rgb(255, 255, 255);"><div style="display: block; border-bottom: 1px solid rgb(230, 230, 230); height: 35px; line-height: 35px; margin: 0px; padding: 0px; overflow: hidden;"><span style="float: left;display: inline;padding-left: 8px;font: 700 14px/35px SimSun;">设置 & 开关</span></div><div style="display: block; position: absolute; top: 36px; width: 100%; height: calc(100% - 36px);"><div style="height: 100%; overflow: auto; padding: 0px 12px; margin: 0px;"><div><label style="display: inline;" title="调速">现在倍速：<input id="speed" type="text"></label></div><div><label style="display: inline;" title="静音"><input id="video_muted" type="checkbox">静音</label></div><div><label style="display: inline;" title="开始/停止"><button id="startstop">开始学习</button></label></div></div></div></div>';
    document.querySelector('body').appendChild(xfk);
    var speed;
    var muted;
    if (unsafeWindow.localStorage.getItem('speed') == null)
    {
        speed = '1';
    }
    else
    {
        speed = unsafeWindow.localStorage.getItem('speed');
    }
    if (unsafeWindow.localStorage.getItem('muted') == null)
    {
        muted = false;
    }
    else
    {
        if (unsafeWindow.localStorage.getItem('muted') == 'true')
        {
            muted = true;
        }
        else
        {
            muted = false
        }
    }
    for (var i = 0;i <= 6;i++) {
        setTimeout(function() {
            if (unsafeWindow.koLearnCourseViewModel == undefined)
            {
                console.log('函数查找失败');
            }
            else
            {
                console.log('函数存在')
            }
        }, 6000);
    }
    document.querySelector('input[id="speed"]').value = speed;
    document.querySelector('input[id="video_muted"]').checked = muted;
    $("#set-auto").click(
        function()
        {
            if (document.querySelector('body').querySelector('span[id="set-auto"]').innerText == '设置\n&\n开关')
            {
                document.querySelector('div[id="set-mune-hide"]').setAttribute('style','position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px;');
                document.querySelector('div[id="set-mune-hide"]').setAttribute('id','set-mune-unhide');
                document.querySelector('body').querySelector('span[id="set-auto"]').innerHTML = '保存<br />or<br />隐藏';
            }
            else
            {
                document.querySelector('div[id="set-mune-unhide"]').setAttribute('style','position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px; display: none;');
                document.querySelector('div[id="set-mune-unhide"]').setAttribute('id','set-mune-hide');
                document.querySelector('body').querySelector('span[id="set-auto"]').innerHTML = '设置<br />&<br />开关'
                unsafeWindow.localStorage.setItem('speed',document.querySelector("input[id='speed']").value);
                unsafeWindow.localStorage.setItem('muted',document.querySelector("input[id='video_muted']").checked);
            }
        }
    );
    $("#startstop").click(
        function()
        {
            unsafeWindow.localStorage.setItem('speed',document.querySelector("input[id='speed']").value);
            unsafeWindow.localStorage.setItem('muted',document.querySelector("input[id='video_muted']").checked);
            if (document.querySelector('button[id="startstop"]').innerHTML == '开始学习')
            {
                unsafeWindow.watch_class = setInterval(function(){
                    if (document.getElementsByClassName('modal-backdrop fade in')[0] != null) {
                        console.log('提示框');
                        if (document.querySelector("[data-bind='text: $root.i18nMsgText().gotIt']") != null) {
                            console.log('视频提示');
                            document.querySelector("[data-bind='text: $root.i18nMsgText().gotIt']").click();
                        }
                        else if (document.querySelector("[data-bind='text: $root.i18nMsgText().confirmLeave']") != null) {
                            console.log('未完成确认');
                            document.querySelector("[data-bind='text: $root.i18nMsgText().confirmLeave']").click();
                        }
                        else if (document.querySelector("[data-bind='text: $root.i18nMsgText().continueStudy']") != null)
                        {
                            console.log('走神提示');
                            document.querySelector("[data-bind='text: $root.i18nMsgText().continueStudy']").click();
                        }
                        else {
                            alert('未知情况，暂无应对方案，请将本页面截屏，返回给作者。');
                            clearInterval(unsafeWindow.watch_class);
                            return (false);
                        }
                    };
                    var data = new Array();
                    for (var i = 0;i < document.querySelectorAll('video').length;i++)
                    {
                        data[i] = new Array();
                        data[i]["video"] = document.querySelectorAll('video')[i];
                    }
                    var counts = 0;
                    for (var count = 0;count < document.querySelectorAll("div[class='video-info'] span").length;count++)
                    {
                        var data_bind = document.querySelectorAll("div[class='video-info'] span")[count].getAttribute('data-bind');
                        if (data_bind == 'text: $root.i18nMessageText().finished')
                        {
                            data[counts]['s'] = true;
                            counts++;
                            console.log(true);
                        }
                        else if (data_bind == 'text: $root.i18nMessageText().viewed' || data_bind == 'text: $root.i18nMessageText().unviewed')
                        {
                            data[counts]['s'] = false;
                            counts++;
                            console.log(false);
                        }
                    }
                    var all = true
                    for (var j = 0;data.length > j;j++)
                    {
                        if (data[j]['s'] == false)
                        {
                            all = false;
                            if (data[j]['video'].paused == true)
                            {
                                data[j]['video'].play();
                                data[j]['video'].muted = muted;
                                data[j]['video'].playbackRate = unsafeWindow.localStorage.getItem('speed');
                                data[j]['video'].muted = unsafeWindow.localStorage.getItem('muted');
                            }
                            else
                            {
                                if (unsafeWindow.localStorage.getItem('muted') == 'true')
                                {
                                    if (data[j]['video'].muted != true)
                                    {
                                        data[j]['video'].muted = true;
                                    }
                                }
                                else
                                {
                                    if (data[j]['video'].muted != false)
                                    {
                                        data[j]['video'].muted = false;
                                    }
                                }
                                if (data[j]['video'].playbackRate != unsafeWindow.localStorage.getItem('speed'))
                                {
                                    data[j]['video'].playbackRate = unsafeWindow.localStorage.getItem('speed');
                                }
                            }
                            break;
                        }
                    }
                    if (all == true)
                    {
                        console.log('下一页');
                        unsafeWindow.koLearnCourseViewModel.goNextPage();
                        for (var k = 0; k < document.querySelectorAll("[data-bind='text: $root.nextPageName()']").length; ++k) {
                            console.log(document.querySelectorAll("[data-bind='text: $root.nextPageName()']")[k].innerHTML);
                            if (document.querySelectorAll("[data-bind='text: $root.nextPageName()']")[k].innerHTML == "没有了") {
                                setTimeout(function(){
                                    try
                                    {
                                        unsafeWindow.koLearnCourseViewModel.goBack();
                                    }
                                    catch (error)
                                    {
                                        console.log(error);
                                    }
                                },1500);
                                clearInterval(unsafeWindow.watch_class);
                                return (true);
                            };
                        };
                    }
                },1500);
                document.querySelector('button[id="startstop"]').innerHTML = '停止学习';
            }
            else
            {
                clearInterval(unsafeWindow.watch_class);
                document.querySelector('button[id="startstop"]').innerHTML = '开始学习';
            }
        }
    );
})();

优学院看视频.user.js
// ==UserScript==
// @name         优学院看视频
// @namespace    https://greasyfork.org/zh-CN/scripts/382033-优学院看视频
// @version      2019.11.30
// @description  可用来看优学院视频而不用手动点击
// @author       Brush-JIM
// @match        https://ua.ulearning.cn/learnCourse/learnCourse.html?*
// @grant        unsafeWindow
// @grant        GM_getValue
// @grant        GM_setValue
// @grant        GM.getValue
// @grant        GM.setValue
// @grant        GM_deleteValue
// @grant        GM.deleteValue
// @grant        GM_xmlhttpRequest
// @grant        GM.xmlhttpRequest
// @run-at       document-start
// @require      https://code.jquery.com/jquery-3.3.1.min.js
// @icon         https://www.ulearning.cn/ulearning/favicon.ico
// @supportURL   https://greasyfork.org/zh-CN/scripts/382033-优学院看视频
// ==/UserScript==

(function() {
    'use strict';
    var ctrl_state=false;
    var learning_state=false;
    unsafeWindow.localStorage.removeItem('failureRecord');
    try{unsafeWindow.document.__defineGetter__('hidden',function(){return false})}catch(e){Object.defineProperty(unsafeWindow.document,'hidden',{get:function(){return false}})};
    try{unsafeWindow.document.__defineGetter__('visibilityState',function(){return 'visible'})}catch(e){Object.defineProperty(unsafeWindow.document,'visibilityState',{get:function(){return 'visible'}})};
    function gm_get(name,defaultValue){if(typeof GM_getValue==='function'){return new Promise((resolve,reject)=>{resolve(GM_getValue(name,defaultValue))})}else{return GM.getValue(name,defaultValue)}};
    function gm_set(name,defaultValue){if(typeof GM_setValue==='function'){GM_setValue(name,defaultValue)}else{GM.setValue(name,defaultValue)}};
    function gm_xml(obj){if(typeof GM_xmlhttpRequest==='fcuntion'){GM_xmlhttpRequest(obj)}else{GM.xmlhttpRequest(obj)}};
    function save_set(){gm_set('speed',window.speed);gm_set('muted',window.muted);gm_set('auto_exit',window.auto_exit)};
    function get_ele_set(){window.speed=$('input[id="speed"]')[0].value;window.muted=$('input[id="video_muted"]')[0].checked;window.auto_exit=$('input[id="exit"]')[0].checked};
    function set_ele(){$('input[id="speed"]')[0].value=window.speed;$('input[id="video_muted"]')[0].checked=window.muted;$('input[id="exit"]')[0].checked=window.auto_exit}
    gm_get('speed').then((speed)=>{
        gm_get('muted').then((muted)=>{
            gm_get('auto_exit').then((auto_exit)=>{
                if(speed==undefined){speed=1}
                if(muted==undefined){muted=false}
                if(auto_exit==undefined){auto_exit=true}
                window.speed=speed;window.muted=muted;window.auto_exit=auto_exit;save_set();
                $(unsafeWindow.document).ready(function(){
                    unsafeWindow.document.body.addEventListener('DOMSubtreeModified',function(){unsafeWindow.$(unsafeWindow).off('beforeunload')} , false);
                    add_ele();
                });
            })
        })
    })
    function learning(){
        if(learning_state==false){
            learning_state=true;
            document.querySelector('button[id="startstop"]').innerHTML = '停止学习';
            var learning_class=function(){
                if(document.getElementsByClassName('modal-backdrop fade in')[0]!=null){
                    var cl_fade=false;
                    if(document.querySelector("[data-bind='text: $root.i18nMsgText().gotIt']")!=null){document.querySelector("[data-bind='text: $root.i18nMsgText().gotIt']").click();cl_fade=true}
                    if(document.querySelector("[data-bind='text: $root.i18nMsgText().confirmLeave']")!=null){
                        document.querySelector("[data-bind='text: $root.i18nMsgText().confirmLeave']").click();cl_fade=true
                    }
                    if(document.querySelector("[data-bind='text: i18nMessageText().nextChapter']")!=null){
                        document.querySelector("[data-bind='text: i18nMessageText().nextChapter']").click();cl_fade=true
                    }
                    if(cl_fade==false){
                        alert('未知情况，请反馈问题。');
                        document.querySelector('button[id="startstop"]').innerHTML = '开始学习';
                        return false
                    }
                    else{
                        if(learning_state==true){
                            setTimeout(learning_class,2000)
                        }
                    }
                }
                else{
                    var data = new Array();
                    var all_finsh=false;
                    for (let i = 0;i < $('video').length;i++)
                    {
                        data[i] = new Array();
                        data[i]['video'] = document.querySelectorAll('video')[i];
                        data[i]['video'].onwaiting=(event)=>{data[i]['onwaiting']=true}
                        data[i]['state']=false;
                        data[i]['ele_tips']=false;
                        data[i]['onwaiting']=false;
                    }
                    function watch_video(){
                        if(all_finsh==true){
                            if(learning_state==true){
                                setTimeout( function () {
                                    unsafeWindow.koLearnCourseViewModel.goNextPage();
                                    for (var k = 0; k < document.querySelectorAll("[data-bind='text: $root.nextPageName()']").length; ++k) {
                                        if (document.querySelectorAll("[data-bind='text: $root.nextPageName()']")[k].innerHTML == "没有了") {
                                            if (window.auto_exit == true)
                                            {
                                                unsafeWindow.koLearnCourseViewModel.goBack();
                                            }
                                            else
                                            {
                                                document.querySelector('button[id="startstop"]').innerHTML = '学习完成（该按钮已不可点击）';
                                                document.querySelector('button[id="startstop"]').disabled = true;
                                            }
                                            return;
                                        }
                                    }
                                    setTimeout(learning_class,2000)
                                },2000)
                            }
                        }
                        else
                        {
                            var counts=0;
                            for(let count=0;count<document.querySelectorAll("div[class='video-info'] span").length;count++){
                                let data_bind = document.querySelectorAll("div[class='video-info'] span")[count].getAttribute('data-bind');
                                if (data_bind == 'text: $root.i18nMessageText().finished')
                                {
                                    data[counts]['ele_tips']=true
                                    if(data[counts]['onwaiting']==false){
                                        if(data[counts]['video'].paused==true){
                                            data[counts]['state']=true
                                        }
                                    }else{
                                        data[counts]['onwaiting']=false
                                    }
                                    counts++;
                                }
                                else if (data_bind == 'text: $root.i18nMessageText().viewed' || data_bind == 'text: $root.i18nMessageText().unviewed'){
                                    counts++;
                                }
                            }
                            var all=true
                            for(let j=0;j<data.length;j++){
                                if(data[j]['ele_tips']==false && data[j]['state']==false){
                                    all=false;
                                    if(data[j]['video'].paused==true){
                                        data[j]['video'].play();
                                        if(data[j]['video'].muted!=window.muted){
                                            data[j]['video'].muted = window.muted;
                                        }
                                        if(data[j]['video'].playbackRate!=window.speed){
                                            data[j]['video'].playbackRate = window.speed;
                                        }
                                    }
                                    else{
                                        if(data[j]['video'].muted!=window.muted){
                                            data[j]['video'].muted = window.muted;
                                        }
                                        if(data[j]['video'].playbackRate!=window.speed){
                                            data[j]['video'].playbackRate = window.speed;
                                        }
                                    }
                                    break;
                                }
                                else if(data[j]['ele_tips']==true && data[j]['state']==false){
                                    all=false
                                }
                            }
                            if(all==true){all_finsh=true}
                            setTimeout(watch_video,2000)
                        }
                    }
                    watch_video();
                }
            }
            learning_class()
        }else{
            learning_state=false;
            document.querySelector('button[id="startstop"]').innerHTML = '开始学习';
        }}
    function ctrl_mune(){
        if(ctrl_state==false){
            ctrl_state=true;
            $('div[id="set-mune"]').attr('style','position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px;');
            $('span[id="set-auto"]')[0].innerHTML = '保存<br />&<br />隐藏';
        }
        else{
            ctrl_state=false;
            $('div[id="set-mune"]').attr('style','position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px; display: none;');
            $('span[id="set-auto"]')[0].innerHTML = '设置<br />&<br />开关'
        }
        get_ele_set()
        save_set()
    }
    function add_ele(){
        try{
            document.getElementById("statModal").parentNode.removeChild(document.getElementById("statModal"));
            document.getElementsByClassName("user-guide")[0].parentNode.removeChild(document.getElementsByClassName("user-guide")[0]);
        }
        catch(e){;}
        var e=document.createElement('div');
        e.setAttribute("style", "width: 60px;height: 74px;overflow: hidden;position: fixed;right: 10px;bottom: 10%;padding: 4px;background-color: #06a9f4;z-index: 9999;border: 1px solid rgb(233, 234, 236);");e.setAttribute("id","set-auto");e.innerHTML='<div style="cursor: pointer; text-align: center; padding: 0px;"><span id="set-auto" style="font-size: 12px;line-height: 16px;color: rgb(255, 255, 255);">设置<br />&<br />开关</span></div>';document.querySelector('body').appendChild(e);
        var e_=document.createElement('div');
        e_.setAttribute("style", "position: fixed;height: 300px;bottom: 10%;z-index: 9999;right: 70px; display: none;");e_.setAttribute("id", "set-mune");e_.innerHTML = '<div style="display: block;overflow: hidden;height: 300px;width: 300px;border: 1px solid rgb(233, 234, 236);background-color: rgb(255, 255, 255);"><div style="display: block; border-bottom: 1px solid rgb(230, 230, 230); height: 35px; line-height: 35px; margin: 0px; padding: 0px; overflow: hidden;"><span style="float: left;display: inline;padding-left: 8px;font: 700 14px/35px SimSun;">设置 & 开关</span></div><div style="display: block; position: absolute; top: 36px; width: 100%; height: calc(100% - 36px);"><div style="height: 100%; overflow: auto; padding: 0px 12px; margin: 0px;"><div><label style="display: inline;" title="调速">现在倍速：<input id="speed" type="number" min="0.5" max="5"></label></div><div><label style="display: inline;" title="静音"><input id="video_muted" type="checkbox">静音</label></div><div><label style="display: inline;" title="退出"><input id="exit" type="checkbox">完成后返回课程目录</label></div><div><label style="display: inline;" title="开始/停止"><button id="startstop">开始学习</button></label></div></div></div></div>';document.querySelector('body').appendChild(e_);
        set_ele();
        $('#set-auto').click(ctrl_mune);
        $('#startstop').click(learning)
    }
})();